[Inference] Use huggingface_hub inference client for TGI adapter

hanouticelina opened the pull request:
### Why this PR? What does it do?
This PR refines the existing TGI integration #52. Specifically:

- Updates the TGI client from the _legacy_ version to the current, maintained inference client (documentation [here](https://huggingface.co/docs/huggingface_hub/v0.24.6/en/package_reference/inference_client#inference)).
- Adds explicit documentation for using TGI with [Hugging Face Inference Endpoints](https://huggingface.co/inference-endpoints/dedicated).

Test Plan
I tested the refined TGI integration with both local endpoint and Hugging Face Inference Endpoints. 

**1. Setup the distribution**
Locally, we need to first start up the TGI container:
```
docker run --rm -it \
     --network host \
    -e HF_TOKEN=$HF_TOKEN \
    -v $HOME/.cache/huggingface:/data \
    ghcr.io/huggingface/text-generation-inference:latest \
    --dtype bfloat16 \
    --model-id meta-llama/Meta-Llama-3.1-8B-Instruct
    --port 5555
```
Then we build a llama stack
`llama stack build local-plus-tgi-inference --name 8b-instruct`

This command creates a conda environment with the necessary dependencies. Then, you'll be prompted for:

- TGI endpoint URL: For local endpoints: http://localhost:5555 and for Hugging Face Inference Endpoints, use the provided endpoint URL.
- Hugging Face Token (if using Hugging Face Inference Endpoints).

**2. Run the server**
` llama stack run local-plus-tgi-inference --name 8b-instruct --port 6001`

**3. Test the distribution**
`python -m llama_toolchain.inference.client localhost 6001`

ashwinb commented:
I don't think the openai-compat layer works correctly with tool calling. I specifically needed to use `generate_stream()` with explicit configuration for stop tokens (<|eom_id|>, <|eot_id|>) without which things don't work. 

Can you show a test where you point `python -m llama_toolchain.agentic_system.client` to this server and show that the brave search tool calling example  (Who's the president of US) works?

hanouticelina commented:
> I don't think the openai-compat layer works correctly with tool calling. I specifically needed to use `generate_stream()` with explicit configuration for stop tokens (<|eom_id|>, <|eot_id|>) without which things don't work.
> 
> Can you show a test where you point `python -m llama_toolchain.agentic_system.client` to this server and show that the brave search tool calling example (Who's the president of US) works?

Hi @ashwinb, you were right, we need full control of the tokens sent to the model which is not easy to get using `InferenceClient.chat_completion()`. I ended up using `huggingface_hub.Inference_Client.text_generation(stream=True)` (documentation [here](https://huggingface.co/docs/huggingface_hub/v0.24.6/en/package_reference/inference_client#huggingface_hub.InferenceClient.text_generation)), which is maintained on our side and has the same behavior as `text_generation.generate_stream()`.

I tested the brave search tool calling with the Server:

1. Setup and build the environnement
```
llama stack build adhoc \
  agentic_system=meta-reference,safety=meta-reference,memory=meta-reference-faiss,inference=remote::tgi \
 --name tgi
```
(works also using the distribution `local-plus-tgi-inference`)
2. Run the server
`llama stack run adhoc --name tgi --port 6001`

_Note_: I unabled Brave Search tool only in the agentic system client.

3. Test the server
`python -m llama_toolchain.agentic_system.client localhost 6001`

Output:
```
User> Search web for who was 44th President of USA?
StepType.inference> brave_search.call(query="44th President of the United States")
StepType.tool_execution> Tool:BuiltinTool.brave_search Args:{'query': '44th President of the United States'}
StepType.tool_execution> Tool:BuiltinTool.brave_search Response:{"query": "44th President of the United States", "top_k": [{"title": "President Barack Obama | Barack Obama Presidential Library", "url": "https://www.obamalibrary.gov/obamas/president-barack-obama", "description": "He accepted the Democratic Party\u2019s nomination at Invesco Stadium in Denver, Colorado on August 28, 2008. On November 4, 2008, Obama became the first African-American to be elected President. He resigned his seat in the U.S. Senate on November 16, 2008. <strong>Barack Obama</strong> was inaugurated as the ...", "type": "search_result"}, {"title": "Barack Obama | The White House", "url": "https://www.whitehouse.gov/about-the-white-house/presidents/barack-obama/", "description": "<strong>Barack Obama</strong> served as the 44th President of the United States. His story is the American story \u2014 values from the heartland, a middle-class upbringing in a strong family, hard work and education as the means of getting ahead, and the conviction that a life so blessed should be lived in service ...", "type": "search_result"}, [{"type": "video_result", "url": "https://www.instagram.com/reel/CzMZbJmObn9/", "title": "Fifteen years ago, on Nov. 4, Barack Obama was elected as ...", "description": "Welcome back to Instagram. Sign in to see what your friends, family and interests have been capturing and sharing around the world."}, {"type": "video_result", "url": "https://video.alexanderstreet.com/watch/the-44th-president-barack-obama?context=channel:barack-obama", "title": "The 44th President (Barack Obama) - Alexander Street, a ...", "description": "You need to enable JavaScript to run this app"}, {"type": "video_result", "url": "https://www.youtube.com/watch?v=iyL7_2-em5k", "title": "Barack Obama for Kids | Learn about the life and contributions ...", "description": "Enjoy the videos and music you love, upload original content, and share it all with friends, family, and the world on YouTube."}, {"type": "video_result", "url": "https://www.britannica.com/video/172743/overview-Barack-Obama", "title": "President of the United States of America Barack Obama | Britannica", "description": "Barack Obama was elected the 44th president of the United States in 2008, becoming the first African American to hold the office. For most of his presidency Obama faced Republican opposition to almost all his proposals. Nevertheless, he was still able to fulfill some of his biggest campaign ..."}, {"type": "video_result", "url": "https://www.youtube.com/watch?v=rvr2g8-5dcE", "title": "The 44th President: In His Own Words - Toughest Day | Special ...", "description": "President Obama reflects on his toughest day in the Presidency and seeing Secret Service cry for the first time. Watch the premiere of The 44th President: In..."}]]}
StepType.shield_call> No Violation
StepType.inference> The 44th President of the United States was Barack Obama.
```

ashwinb commented on llama_toolchain/inference/adapters/tgi/tgi.py:
```diff
@@ -4,80 +4,51 @@
-            import traceback
+    @property
+    def client(self) -> InferenceClient:
+        return InferenceClient(base_url=self.config.url, token=self.config.api_token)
```
Just want to make sure this works if no api_token is specified for the local TGI container

hanouticelina replied:
Hi @ashwinb, The default value of `api_token` is set to an empty string when it's not specified and it's working fine with a local TGI container. However, I changed the default value to `None` to be able to retrieve the token from cache or local , works fine also with a local endpoint.

ashwinb commented on llama_toolchain/inference/adapters/tgi/tgi.py:
```diff
@@ -4,80 +4,51 @@
 
-            traceback.print_exc()
-            raise RuntimeError("Could not connect to TGI server") from e
+    async def initialize(self) -> None:
```
could you move code here so we know at initialization time if the TGI container one is trying to connect to is UP or not? the previous code did this. Also -- we do need to check if the model that this container is serving is supported or not by that container. We are also trying to make sure this works when people deploy TGI containers by themselves (and it is not _just_ the dedicated inference endpoint)

ashwinb commented on llama_toolchain/inference/adapters/tgi/tgi.py:
```diff
@@ -89,62 +60,58 @@ def get_chat_options(self, request: ChatCompletionRequest) -> dict:
         model_input = self.formatter.encode_dialog_prompt(messages)
         prompt = self.tokenizer.decode(model_input.tokens)
+
+        model_info = self.client.get_endpoint_info(model=self.config.url)
```
basically move this to the `initialize()` function

ashwinb reviewed (changes requested):
This is awesome, thank you so much. Just a few comments.

hanouticelina commented:
Hello @ashwinb, thanks for your review! I updated the Adapter based on your comments. I split the TGI adapter for Local endpoints and Hugging Face Inference Endpoints due to differences in retrieving the model_id and other information about the model:

- Local TGI deployments: We can directly use `InferenceClient.get_endpoint_info()` to get the `model_id` and `max_total_tokens`.

- Hugging Face Inference Endpoints: The inference URL doesn't provide the `model_id`. Users need to provide their HF namespace (username or organization) and endpoint name. With these, we retrieve both the inference URL and model information (`model_id` and `max_total_tokens`).

I've tested both adapters for inference and agentic function calling. Please let me know if you have any suggestions for further improvements!

Wauplin commented on llama_toolchain/inference/adapters/tgi/__init__.py:
```diff
@@ -4,12 +4,21 @@
+async def get_adapter_impl(config: TGIImplConfig, _deps):
+    assert isinstance(config, TGIImplConfig), f"Unexpected config type: {type(config)}"
+
+    if config.is_local_tgi():
+        impl = TGIAdapter(config)
```
(small suggestion)

```suggestion
    if config.url is not None:
        impl = TGIAdapter(config)
```

I would only rely on the fact that `config.url` is not None and not the localhost thingy. Since TGI can be deployed anywhere (locally, on an EC2 instance, behind a proxy, etc.), we should not assume what the URL should look like.

hanouticelina replied:
indeed! thanks

ashwinb commented on llama_toolchain/inference/adapters/tgi/config.py:
```diff
@@ -0,0 +1,34 @@
+
+from typing import Optional
+
+from huggingface_hub import HfApi
```
we want the config files to be extremely lightweight so one can use them outside of the larger containers. as such they should only have datatypes and no "behavior" or functions at all. I suggest you move at least the `get_namespace()` method out into the impl where it needs to be used so this dependency is not here.

hanouticelina replied:
Fixed ✅

ashwinb reviewed (approved):
Looks good. Thanks once again.

I do have one important comment about the `get_namespace()` method pulling in the `huggingface_hub` dep into the config file. Please move that helper method outside as a free floating function outside into the impl file.
